presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,268 @@
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.lang.Math.*;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+public class ShardOrganizationManager
+{
+    private static final Logger log = Logger.get(ShardOrganizationManager.class);
+
+    private final ScheduledExecutorService discoveryService = newScheduledThreadPool(1, daemonThreadsNamed("shard-organization-discovery"));
+    private final AtomicBoolean started = new AtomicBoolean();
+    private final AtomicBoolean shutdown = new AtomicBoolean();
+
+    private final IDBI dbi;
+    private final MetadataDao metadataDao;
+    private final ShardOrganizerDao shardOrganizerDao;
+    private final ShardManager shardManager;
+
+    private final boolean enabled;
+    private final Duration discoveryInterval;
+
+    private final String currentNodeIdentifier;
+    private final ShardOrganizer organizer;
+
+    @Inject
+    public ShardOrganizationManager(@ForMetadata IDBI dbi, NodeManager nodeManager, ShardManager shardManager, ShardOrganizer organizer, StorageManagerConfig config)
+    {
+        this(dbi,
+                nodeManager.getCurrentNode().getNodeIdentifier(),
+                shardManager,
+                organizer,
+                config.isOrganizationEnabled(),
+                config.getOrganizationInterval());
+    }
+
+    public ShardOrganizationManager(
+            IDBI dbi,
+            String currentNodeIdentifier,
+            ShardManager shardManager,
+            ShardOrganizer organizer,
+            boolean enabled,
+            Duration discoveryInterval)
+    {
+        this.dbi = requireNonNull(dbi, "dbi is null");
+        this.metadataDao = onDemandDao(dbi, MetadataDao.class);
+        this.shardOrganizerDao = onDemandDao(dbi, ShardOrganizerDao.class);
+
+        this.organizer = requireNonNull(organizer, "organizer is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.currentNodeIdentifier = requireNonNull(currentNodeIdentifier, "currentNodeIdentifier is null");
+
+        this.enabled = enabled;
+        this.discoveryInterval = requireNonNull(discoveryInterval, "discoveryInterval is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!enabled) {
+            return;
+        }
+
+        if (!started.getAndSet(true)) {
+            startDiscovery();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        discoveryService.shutdown();
+    }
+
+    private void startDiscovery()
+    {
+        long startDelay = getStartDelay();
+
+        discoveryService.scheduleWithFixedDelay(() -> {
```

Comment:
Rather than running discovery on a schedule, on startup it should lookup the last run time and compute the delay, then schedule the initial run based on that delay.

The job submission should run in a loop until it succeeds, with some sleep after a failure, rather than waiting for an entire scheduled period.